cpp/include/algorithms.hpp

```diff
@@ -0,0 +1,61 @@
+ * @Param[in] alpha                  The damping factor alpha represents the probability to follow an outgoing edge, standard value is 0.85.
+                                     Thus, 1.0-alpha is the probability to “teleport” to a random vertex. Alpha should be greater than 0.0 and strictly lower than 1.0.
+ *                                   The initial guess must not be the vector of 0s. Any value other than 1 or 0 is treated as an invalid value.
+ * @Param[in] pagerank               Array of size V. Should contain the initial guess if has_guess=true. In this case the initial guess cannot be the vector of 0s. Memory is provided and owned by the caller.
+ * @Param[in] personalization_subset_size (optional) The number of vertices for to personalize. Initialized to 0 by default.
+ * @Param[in] personalization_subset (optional) Array of size personalization_subset_size containing vertices for running personalized pagerank. Initialized to nullptr by default. Memory is provided and owned by the caller.
+ * @Param[in] personalization_values (optional) Array of size personalization_subset_size containing values associated with personalization_subset vertices. Initialized to nullptr by default. Memory is provided and owned by the caller.
+ * @Param[in] tolerance              Set the tolerance the approximation, this parameter should be a small magnitude value.
+ *                                   The lower the tolerance the better the approximation. If this value is 0.0f, cuGRAPH will use the default value which is 1.0E-5.
+ *                                   Setting too small a tolerance can lead to non-convergence due to numerical roundoff. Usually values between 0.01 and 0.00001 are acceptable.
+ * @Param[in] max_iter               (optional) The maximum number of iterations before an answer is returned. This can be used to limit the execution time and do an early exit before the solver reaches the convergence tolerance.
+ *                                   If this value is lower or equal to 0 cuGRAPH will use the default value, which is 500.
+  * @Param[in] has_guess             (optional) This parameter is used to notify cuGRAPH if it should use a user-provided initial guess. False means the user does not have a guess, in this case cuGRAPH will use a uniform vector set to 1/V.
+ *                                   If the value is True, cuGRAPH will read the pagerank parameter and use this as an initial guess.
+ * @Param[out] *pagerank             The PageRank : pagerank[i] is the PageRank of vertex i. Memory remains provided and owned by the caller.
+ *
+ * @throws                           cugraph::logic_error with a custom message when an error occurs.
+ */
+/* ----------------------------------------------------------------------------*/
+template <typename VT, typename WT>
+void pagerank(experimental::GraphCSC<VT,WT> const &graph,
```

Comment:
Should we expose the graph structure in the algo API? We could just accept a "Graph" so that we can change the algorithm implementation without touching the API? Otherwise, we will have an entry point for each implemented format.

Reply:
Exposing the graph structure in the algorithm API allows us to check the validity at compile time.  If you tried using a GraphCOO to call pagerank then you wouldn't be able to compile.

When we have an analytic that can run off of different types of graphs, we end up with a bunch of if statements that  switch us into the correct configuration.  These are, IMO, simpler if you just have two entry points.  They can still call common underlying functions to avoid duplicating code.

Reply:
While I acknowledge the benefits of compile-time checks, multiplying the number of specific API entries has side costs in terms of documentation, testing, and user-friendliness. In general, I tend to prefer (a reasonable amount of) abstractions as a way of future-proofing the API and controlling maintenance. 

@seunghwak @BradReesWork any preference here?

Reply:
I guess the question is how often do we expect to have algorithms that have more than one implementation?  Do we expect to have COO and CSR implementations of many of these algorithms?

Reply:
I think we should better abstract this out. Instead of algorithms directly accessing graph internals,

We should have something like

graph.vertices_begin(), graph.vertices_end(), <= for the entire set of vertices
graph.edges_begin(), graph.edges_end(), <= for the entire set of edges
graph.edges_begin(v), graph.edges_end(v), <= for the edges of vertex v
graph.(incoming or)outgoing_edges_begin(v), graph.(incoming_or)outgoing_edges_end(v) <= for the (incoming or)outgoing edges of vertex v

This way, we can separate graph data structure internals from algorithm developments and the same algorithm will work whether we have static CSR or dynamically updatable CSR or other data types.

Reply:
We should not write new algorithms for every new graph representation.

If I bring a quote from "A tour of C++ by Bjarne Stroustrup (the inventor of C++)"

"An algorithm operates on its data through iterators and knows nothing about the container in which the elements are stored. Conversely, a container knows nothing about the algorithms operating on its elements; all it does is to supply iterators upon request (e.g., begin() and end()). This model of separation between data storage and algorithm delivers very general and flexible software"

I think the recent libcudf porting achieved this goal, and we should follow this as well.

Reply:
And vertex.begin() may be better than vertices_begin()... need more thinking about the exact API, but I think this is the direction we should move.

Reply:
I would agree to the principal here... with the following caveat.  cugraph (at least as it exists today) is largely about integrating a collection of disparate graph libraries (nvgraph, hornet, gunrock, some homegrown analytics).  As long as cugraph integrates libraries of analytics developed by different developers, we need to have an interface that will be common among them.  At this point, I think the lowest common denominator is device pointers to CSR structures.

Unless we plan on either defining a graph API and requiring hornet and gunrock (and any future collaborators we work with) to comply to that API - and significantly rewrite all of the current graph analytics, I think we're limited to a data structure that is like this.

My thought we be to use GraphCOO, GraphCSR and GraphCSC as basic lower level data structures and then writing a higher level abstraction based upon them to provide iterators for new analytics.

Reply:
Yes, you're right.

So if we view this PR as a temporary solution to maintain cugraph functional when libcudf dumps gdf_column, I'm fine with any legitimate solution. Better not invest too much on a temporary solution.

In a long run, having nvgraph, hornet, gunrock, and actual cugraph all under cugraph is not a scalable solution. It will inevitably become a maintainability nightmare as cuGraph gets bigger & bigger. But I guess this is a much bigger work (requiring the Great libcugraph port) something way beyond this PR.